core/src/main/scala/org/apache/spark/RecoverCacheShutdown.scala

```diff
@@ -0,0 +1,234 @@
+   * otherwise kill the executors
+   *
+   * @param execIds the executors to check
+   */
+  private def checkForReplicableBlocks(execIds: Seq[String]) = state.getBlocks(execIds).foreach {
+    case (executorId, NoMoreBlocks) => state.killExecutor(executorId)
+    case (executorId, NotEnoughMemory) => state.killExecutor(executorId)
+    case (executorId, HasCachedBlocks) => replicateBlocks(executorId)
+  }
+
+  /**
+   * Replicate one cached block on an executor. If there are more, repeat. If there are none, check
+   * with the block manager master again. If there is an error, go ahead and kill executor.
+   *
+   * @param execId the executor to save a block one
+   */
+  private def replicateBlocks(execId: String): Unit = {
+    val (response, blockId) = state.replicateFirstBlock(execId)
+    response.onComplete {
+      case scala.util.Success(true) =>
+        logTrace(s"Finished replicating block ${blockId.getOrElse("unknown")} on exec $execId.")
+        replicateBlocks(execId)
+      case scala.util.Success(false) => checkForReplicableBlocks(Seq(execId))
+      case Failure(f) =>
+        logWarning(s"Error trying to replicate block ${blockId.getOrElse("unknown")}.", f)
+        state.killExecutor(execId)
+    }
+  }
+}
+
+private object RecoverCacheShutdown {
+  def apply(eam: ExecutorAllocationManager, conf: SparkConf): RecoverCacheShutdown = {
+    val bmme = SparkEnv.get.blockManager.master.driverEndpoint
+    val state = new RecoverCacheShutdownState(bmme, eam, conf)
+    new RecoverCacheShutdown(state, conf)
+  }
+}
+
+/**
+ * Private class that holds state for all the executors being shutdown.
+ * @param blockManagerMasterEndpoint blockManagerMasterEndpoint
+ * @param executorAllocationManager ExecutorAllocationManager
+ * @param conf spark conf
+ */
+final private class RecoverCacheShutdownState(
+   blockManagerMasterEndpoint: RpcEndpointRef,
+   executorAllocationManager: ExecutorAllocationManager,
+   conf: SparkConf
+ ) extends Logging {
+
+  type ExecMap[T] = mutable.Map[String, T]
+
+  private val forceKillAfterS =
+    conf.getTimeAsSeconds("spark.dynamicAllocation.recoverCachedData.timeout", "120s")
+  private val scheduler =
+    ThreadUtils.newDaemonSingleThreadScheduledExecutor("recover-cache-shutdown-timers")
+
+  private val blocksToSave: ExecMap[mutable.PriorityQueue[RDDBlockId]] = new mutable.HashMap
+  private val savedBlocks: ExecMap[mutable.HashSet[RDDBlockId]] = new mutable.HashMap
+  private val killTimers: ExecMap[ScheduledFuture[_]] = new mutable.HashMap
+
+  /**
+   * Query Block Manager Master for cached blocks.
+   * @param execIds the executors to query
+   * @return a map of executorId to its replication state.
+   */
+  def getBlocks(execIds: Seq[String]): Map[String, ExecutorReplicationState] = synchronized {
+    logDebug(s"Get all RDD blocks for executors: ${execIds.mkString(", ")}.")
+    execIds.map { id =>
+      if (isThereEnoughMemory(id)) {
```

Comment:
If I understand things correctly, shouldn't you be considering all executors being killed in this call? Otherwise won't you be counting the memory available at executors that soon will be killed?

e.g. if you're killing 5 executors, when you call this for the first one, it seems to me like that would still consider the memory of the other 4 as available.

Feels like you should calculate the available memory before iterating over the executors; that could help with fixing the above issue, and would also be cheaper (single BM call vs. one per executor being killed).

Reply:
I rewrote this code and simplified it. I don't call GetMemoryStatus for every executor anymore.